topdown/eval.go

```diff
@@ -792,6 +793,7 @@ func (e *eval) evalCall(terms []*ast.Term, iter unifyIterator) error {
 		Cancel:                 e.cancel,
 		Runtime:                e.runtime,
 		Cache:                  e.builtinCache,
+		NDCache:                e.ndBuiltinCache,
```

Comment:
I'd have expected the NDCache to be a per-query thing that doesn't get reused. Do when a decision is logged, we can throw it away. If we share it among queries, I'm not sure how we'd figure out which of its values we need to include in a decision logs entry. 🤔 Of course, it can still be part of this struct and be reset. Sorry, this is a premature comment 😅

Reply:
I think you may be right about this. :thinking: I had shied away from using that cache initially, because this change could greatly increase pressure on that cache, and because there was minimal documentation around the existing builtin caching tooling. However! It may be a good fit after all. Assuming I can tie into it correctly, I could take advantage of existing caching infrastructure that others have worked out the plumbing for.

Reply:
> I'd have expected the NDCache to be a per-query thing that doesn't get reused.

Will it be reused? I don't know of another reasonable way to propagate a cache down to the individual builtins, except to provide it in the `BuiltinContext` somehow, which would have to come from, and be managed by the evaluator, or something further up. :thinking:

Reply:
Maybe this is a bad idea, but perhaps passing down the NDCache from the top, at eval time, might be the cleanest way to extract that information, without polluting the normal inter-query cache. :thinking: The default would be that the member in the eval context for the NDCache will be left nil, and thus the default will be to *not* collect that information. This also moves the responsibility for extracting NDCache info as close to the edges as possible. (It'd probably involve a `.WithNDCache(...)` settings call, I think.)

Reply:
> without polluting the normal inter-query cache.

Do we need to worry, though? I'd think that the intra-query and the ND cache would carry exactly the same information at the end of a query eval -- but it's difficult to pluck it out of the intra-query cache. So maybe the functionality could be combined...? 

I it would mean that some of the caching logic in `http.send` needs to be revisited: I think right now, it'll use either the inter- or the intra-query cache (exclusively).